File: asr-service/benchmarks/performance_benchmark.py
from __future__ import annotations

import json
import platform
import statistics
import sys
import tempfile
import time
from pathlib import Path
from typing import Callable

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT))

from echonote_asr.diarization import (  # noqa: E402
    SpeakerInterval,
    assign_speakers_to_turns,
    merge_adjacent_turns,
)
from echonote_asr.schemas import TranscriptTurn  # noqa: E402
from echonote_asr.text_sanitizer import sanitize_transcript_text  # noqa: E402


def main() -> None:
    turns = [
        TranscriptTurn(
            id=f"turn-{index + 1}",
            text=f"Turn {index + 1} discusses performance.",
            speaker=None,
            started_at_ms=index * 3_000,
            ended_at_ms=(index + 1) * 3_000,
            confidence=0.9,
        )
        for index in range(2_000)
    ]
    intervals = [
        SpeakerInterval(
            speaker=f"speaker-{index % 4}",
            started_at_ms=index * 1_500,
            ended_at_ms=(index + 1) * 1_500,
        )
        for index in range(4_000)
    ]
    assigned_turns, _ = assign_speakers_to_turns(turns, intervals)
    pathological_text = "EchoNote performance regression sentence. " * 500
    wav_payload = bytes(44 + (16_000 * 15 * 2))

    def legacy_temp_write() -> None:
        with tempfile.TemporaryDirectory(prefix="echonote-asr-legacy-") as tmp_dir:
            Path(tmp_dir, "chunk.wav").write_bytes(wav_payload)

    reusable_workspace = tempfile.TemporaryDirectory(prefix="echonote-asr-reused-")
    reusable_wav_path = Path(reusable_workspace.name, "chunk.wav")

    def reused_temp_write() -> None:
        reusable_wav_path.write_bytes(wav_payload)
        reusable_wav_path.unlink()

    results = [
        benchmark(
            "diarization.assign_speakers.2k_turns_4k_intervals",
            5,
            lambda: assign_speakers_to_turns(turns, intervals),
        ),
        benchmark(
            "diarization.merge_adjacent.2k_turns",
            20,
            lambda: merge_adjacent_turns(assigned_turns),
        ),
        benchmark(
            "transcript.sanitize.pathological",
            50,
            lambda: sanitize_transcript_text(pathological_text),
        ),
        benchmark("asr.temp_io.legacy_workspace", 50, legacy_temp_write),
        benchmark("asr.temp_io.reused_workspace", 50, reused_temp_write),
    ]
    reusable_workspace.cleanup()

    print(
        json.dumps(
            {
                "runtime": {
                    "python": platform.python_version(),
                    "platform": platform.platform(),
                    "machine": platform.machine(),
                },
                "inputs": {
                    "turns": len(turns),
                    "intervals": len(intervals),
                },
                "results": results,
            },
            indent=2,
        )
    )


def benchmark(name: str, iterations: int, operation: Callable[[], object]) -> dict[str, object]:
    for _ in range(2):
        operation()

    durations_ms: list[float] = []
    for _ in range(iterations):
        started_at = time.perf_counter()
        operation()
        durations_ms.append((time.perf_counter() - started_at) * 1_000)

    ordered = sorted(durations_ms)
    p95_index = min(len(ordered) - 1, max(0, int(len(ordered) * 0.95)))
    return {
        "name": name,
        "iterations": iterations,
        "medianMs": round(statistics.median(ordered), 3),
        "p95Ms": round(ordered[p95_index], 3),
        "minMs": round(ordered[0], 3),
        "maxMs": round(ordered[-1], 3),
    }


if __name__ == "__main__":
    main()

File: plugin/benchmarks/performance-benchmark.ts
import { performance } from "node:perf_hooks";
import type { TranscriptTurn } from "../src/asr/asr-types";
import { MeetingAudioSpool } from "../src/audio/meeting-audio-spool";
import { concatWavFiles } from "../src/audio/wav-encoder";
import {
  extractTranscript,
  isEchoNoteMeetingNote,
  replaceSummarySections
} from "../src/meeting/markdown-sections";
import { formatTranscriptTurns } from "../src/meeting/transcript-markdown";
import { sanitizeTranscriptText } from "../src/meeting/transcript-sanitizer";

type BenchmarkResult = {
  name: string;
  iterations: number;
  medianMs: number;
  p95Ms: number;
  minMs: number;
  maxMs: number;
};

type AudioSpoolBenchmarkResult = {
  name: string;
  minutes: number;
  chunks: number;
  storageMode: "memory" | "disk";
  appendMs: number;
  assemblyMs: number;
  outputMiB: number;
  retainedArrayBufferDeltaMiB: number;
};

const transcriptLines = Array.from({ length: 20_000 }, (_, index) =>
  `[${formatOffset(index * 3_000)}] Speaker ${(index % 4) + 1}: 第 ${index + 1} 条性能基准转录，讨论产品、工程和后续行动。`
);
const transcript = transcriptLines.join("\n");
const meetingNote = `<!-- echonote-meeting -->
# Performance baseline

## Summary

_Pending._

## Decisions

_Pending._

## Action Items

_Pending._

## Key Points

_Pending._

## Open Questions

_Pending._

## Transcript

${transcript}
`;
const turns: TranscriptTurn[] = Array.from({ length: 10_000 }, (_, index) => ({
  id: `turn-${index + 1}`,
  text: `Turn ${index + 1} discusses performance and a concrete follow-up action.`,
  speaker: `Speaker ${(index % 4) + 1}`,
  started_at_ms: index * 3_000,
  ended_at_ms: (index + 1) * 3_000,
  confidence: 0.9
}));
const pathologicalTranscript = "EchoNote performance regression sentence. ".repeat(500);
const wavChunks = Array.from({ length: 40 }, () => new ArrayBuffer(44 + (16_000 * 15 * 2)));

void main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

async function main(): Promise<void> {
  const results: BenchmarkResult[] = [
    runBenchmark("markdown.extract_transcript.20k_lines", 20, () => extractTranscript(meetingNote)),
    runBenchmark("markdown.verify_meeting.20k_lines", 20, () => isEchoNoteMeetingNote(meetingNote)),
    runBenchmark("markdown.replace_summary.20k_lines", 20, () => replaceSummarySections(meetingNote, {
      Summary: "Performance summary",
      Decisions: "- Keep benchmark gates",
      "Action Items": "- [ ] Optimize hot paths",
      "Key Points": "- Measure before changing",
      "Open Questions": "- None"
    })),
    runBenchmark("transcript.format.10k_turns", 10, () => formatTranscriptTurns(turns, true, "")),
    runBenchmark("transcript.sanitize.pathological", 50, () => sanitizeTranscriptText(pathologicalTranscript)),
    runBenchmark("wav.concat.10_minutes", 5, () => concatWavFiles(wavChunks))
  ];
  const audioSpool = [];
  for (const minutes of [10, 30, 60]) {
    audioSpool.push(await runAudioSpoolBenchmark(minutes));
  }

  console.log(JSON.stringify({
    runtime: {
      node: process.version,
      platform: process.platform,
      arch: process.arch
    },
    inputs: {
      noteChars: meetingNote.length,
      transcriptLines: transcriptLines.length,
      transcriptTurns: turns.length,
      wavChunks: wavChunks.length,
      wavMinutes: 10
    },
    results,
    audioSpool
  }, null, 2));
}

async function runAudioSpoolBenchmark(minutes: number): Promise<AudioSpoolBenchmarkResult> {
  const chunks = minutes * 4;
  const chunk = new ArrayBuffer(44 + (16_000 * 15 * 2));
  const spool = new MeetingAudioSpool();
  global.gc?.();
  const initialArrayBuffers = process.memoryUsage().arrayBuffers;
  const appendStartedAt = performance.now();
  for (let index = 0; index < chunks; index += 1) {
    await spool.append(chunk);
  }
  const appendMs = performance.now() - appendStartedAt;
  const storageMode = spool.storageMode;
  const assemblyStartedAt = performance.now();
  const wav = await spool.toWav();
  const assemblyMs = performance.now() - assemblyStartedAt;
  const retainedArrayBufferDelta = Math.max(0, process.memoryUsage().arrayBuffers - initialArrayBuffers);
  const result = {
    name: `audio_spool.stop.${minutes}_minutes`,
    minutes,
    chunks,
    storageMode,
    appendMs: round(appendMs),
    assemblyMs: round(assemblyMs),
    outputMiB: round((wav?.byteLength ?? 0) / (1024 * 1024)),
    retainedArrayBufferDeltaMiB: round(retainedArrayBufferDelta / (1024 * 1024))
  };
  await spool.dispose();
  return result;
}

function runBenchmark(name: string, iterations: number, operation: () => unknown): BenchmarkResult {
  for (let index = 0; index < 3; index += 1) {
    operation();
  }
  global.gc?.();

  const durations: number[] = [];
  for (let index = 0; index < iterations; index += 1) {
    const startedAt = performance.now();
    operation();
    durations.push(performance.now() - startedAt);
  }
  durations.sort((left, right) => left - right);

  return {
    name,
    iterations,
    medianMs: round(percentile(durations, 0.5)),
    p95Ms: round(percentile(durations, 0.95)),
    minMs: round(durations[0]),
    maxMs: round(durations[durations.length - 1])
  };
}

function percentile(values: number[], ratio: number): number {
  const index = Math.min(values.length - 1, Math.max(0, Math.ceil(values.length * ratio) - 1));
  return values[index];
}

function round(value: number): number {
  return Math.round(value * 1_000) / 1_000;
}

function formatOffset(offsetMs: number): string {
  const totalSeconds = Math.floor(offsetMs / 1_000);
  const hours = Math.floor(totalSeconds / 3_600);
  const minutes = Math.floor((totalSeconds % 3_600) / 60);
  const seconds = totalSeconds % 60;
  return [hours, minutes, seconds].map((value) => String(value).padStart(2, "0")).join(":");
}
